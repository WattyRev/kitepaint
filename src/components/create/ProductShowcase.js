import React from "react";
import styled from "styled-components";
import { Link } from "react-router-dom";
import productShape from "../../models/product";
import { H3 } from "../../theme";

/**
 * Provides styling for the showcase
 */
export const StyleWrapper = styled.div`
  flex-basis: 500px;
  margin: 8px;
  flex-shrink: 1;
  border: 1px solid ${props => props.theme.colors.gray};
  border-radius: 4px;
  ${props => props.theme.patterns.transparencyBackground};
  cursor: pointer;
  position: relative;
  display: flex;
  padding: 36px 8px 8px;
  flex-wrap: wrap;
  align-items: middle;
  justify-content: space-around;

  .product-heading {
    background: ${props => props.theme.colors.white};
    border-bottom: 1px solid ${props => props.theme.colors.gray};
    text-align: center;
    padding: 4px 0;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    border-top-left-radius: 4px;
    border-top-right-radius: 4px;
  }
  .preview {
    width: 50%;
    flex-grow: 1;
    padding: 8px;
    box-sizing: border-box;
    transition: 0.6s transform;

    &.count-2 {
      width: 100%;
    }
    > svg {
      height: 100%;
      width: 100%;
    }
  }
  &:hover {
    .preview {
      transform: translateY(-15px);
    }
  }
`;

/**
 * A preview of a specific product, which links to the create page for that product.
 */
const ProductShowcase = ({ product }) => (
  <StyleWrapper as={Link} to={`/create/${product.id}`}>
    <H3 className="product-heading">{product.name}</H3>
    {product.variations.map(variation => (
      <div
        className={`preview count-${product.variations.length}`}
        key={variation.name}
        dangerouslySetInnerHTML={{ __html: variation.svg }}
      />
    ))}
  </StyleWrapper>
);

ProductShowcase.propTypes = {
  /**
   * The product being showcased
   */
  product: productShape.isRequired
};

export default ProductShowcase;
