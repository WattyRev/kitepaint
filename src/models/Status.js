import PropTypes from "prop-types";

/**
 * Status can be applied to any model to indicate it's public visibility.
 */
export const statusProp = PropTypes.oneOf(["0", "1", "2"]);
const Status = {
  "0": "Private",
  "1": "Unlisted",
  "2": "Public",
  PRIVATE: "0",
  UNLISTED: "1",
  PUBLIC: "2"
};
export default Status;
