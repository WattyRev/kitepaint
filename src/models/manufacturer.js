import PropTypes from "prop-types";

/**
 * A manufacturer is a company that manufactures kites and owns one or more products.
 */
const manufacturerState = PropTypes.shape({
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  logo: PropTypes.string.isRequired,
  website: PropTypes.string
});

export default manufacturerState;

/**
 * A mock manufacturer used for testing
 * @type {Object}
 */
const getMockManufacturer = () => ({
  id: "abc",
  name: "Krazy Kites",
  logo: "abc.jpg",
  website: "http://krazykites.com"
});
export { getMockManufacturer };
