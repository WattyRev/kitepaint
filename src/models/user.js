import PropTypes from "prop-types";

/**
 * A user is an individual user account. It contains limited details that are available publically.
 */
const userShape = PropTypes.shape({
  username: PropTypes.string.isRequired,
  loginid: PropTypes.string.isRequired
});

export default userShape;

/**
 * A mock user used for testing.
 */
const getMockUser = () => ({
  username: "FlyFucker",
  loginid: "abc"
});

export { getMockUser };
