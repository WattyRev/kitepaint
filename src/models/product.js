import PropTypes from "prop-types";
import { getMockManufacturer } from "./manufacturer";
import Status, { statusProp } from "./status";

/**
 * A product is a kite that can be customized to create a design.
 */
const productShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  embed: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  notes: PropTypes.arrayOf(PropTypes.string),
  manufacturer: PropTypes.string.isRequired,
  status: statusProp.isRequired,
  url: PropTypes.string,
  colors: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      color: PropTypes.string.isRequired
    })
  ),
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired
    })
  )
});

export default productShape;

/**
 * A mock product used for testing
 * @type {Object}
 */
const getMockProduct = () => ({
  id: "product-1",
  embed: "krazykites.com",
  name: "Krazy Kite",
  manufacturer: getMockManufacturer().id,
  url: "http://krazykites.com/krazy-kite",
  colors: [
    {
      name: "red",
      color: "#ff0000"
    },
    {
      name: "black",
      color: "#000000"
    }
  ],
  variations: [
    {
      name: "Standard",
      svg: "<div>Kool Kite</div>"
    }
  ],
  status: Status.PUBLIC
});

export { getMockProduct };
