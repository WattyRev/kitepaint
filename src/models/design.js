import PropTypes from "prop-types";
import { getMockProduct } from "./product";
import Status, { statusProp } from "./status";

/**
 * A design is a user-created entity. It represetns a saved design of a particular product.
 */
const designShape = PropTypes.shape({
  id: PropTypes.string.isRequired,
  // A string representing the created date
  created: PropTypes.string.isRequired,
  // A string representing the updated date
  updated: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  // The ID of the cooresponding product
  product: PropTypes.string.isRequired,
  // The ID of the user that created the design
  user: PropTypes.string,
  variations: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      svg: PropTypes.string.isRequired,
      primary: PropTypes.bool.isRequired
    })
  ),
  // The status of the cooresponding product
  productStatus: statusProp.isRequired,
  // The status of the design
  status: statusProp.isRequired
});

export default designShape;

/**
 * A mock design used for testing
 * @type {Object}
 */
const getMockDesign = () => ({
  id: "design-1",
  created: "01/01/2018",
  updated: "01/02/2018",
  name: "picasso",
  product: getMockProduct().id,
  variations: [
    {
      name: "Standard",
      svg: "<div>picasso standard</div>",
      primary: true
    }
  ],
  status: Status.PRIVATE,
  productStatus: Status.PUBLIC
});
export { getMockDesign };
