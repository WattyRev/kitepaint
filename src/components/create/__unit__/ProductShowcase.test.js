import React from "react";
import { shallow, mount } from "enzyme";
import { getMockProduct } from "../../../models/product";
import Theme from "../../../theme";
import ProductShowcase, { StyleWrapper } from "../ProductShowcase";

describe("ProductShowcase", () => {
  let defaultProps;
  beforeEach(() => {
    defaultProps = {
      product: getMockProduct()
    };
  });
  it("renders", () => {
    shallow(<ProductShowcase {...defaultProps} />);
  });
  describe("StyleWrapper", () => {
    it("renders", () => {
      expect.assertions(1);
      const wrapper = mount(<StyleWrapper theme={Theme} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
