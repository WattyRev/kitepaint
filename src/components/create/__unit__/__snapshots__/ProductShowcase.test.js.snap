// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductShowcase StyleWrapper renders 1`] = `
.c0 {
  -webkit-flex-basis: 500px;
  -ms-flex-preferred-size: 500px;
  flex-basis: 500px;
  margin: 8px;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  border: 1px solid #AAAAAA;
  border-radius: 4px;
  background-image: linear-gradient(45deg,#AAAAAA 25%,transparent 25%), linear-gradient(-45deg,#AAAAAA 25%,transparent 25%), linear-gradient(45deg,transparent 75%,#AAAAAA 75%), linear-gradient(-45deg,transparent 75%,#AAAAAA 75%);
  background-color: #5e5e5e;
  background-size: 20px 20px;
  background-position: 0 0,0 10px,10px -10px,-10px 0px;
  cursor: pointer;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 36px 8px 8px;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: middle;
  -webkit-box-align: middle;
  -ms-flex-align: middle;
  align-items: middle;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c0 .product-heading {
  background: #FFFFFF;
  border-bottom: 1px solid #AAAAAA;
  text-align: center;
  padding: 4px 0;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.c0 .preview {
  width: 50%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 8px;
  box-sizing: border-box;
  -webkit-transition: 0.6s -webkit-transform;
  -webkit-transition: 0.6s transform;
  transition: 0.6s transform;
}

.c0 .preview.count-2 {
  width: 100%;
}

.c0 .preview > svg {
  height: 100%;
  width: 100%;
}

.c0:hover .preview {
  -webkit-transform: translateY(-15px);
  -ms-transform: translateY(-15px);
  transform: translateY(-15px);
}

<styled.div
  theme={
    Object {
      "colors": Object {
        "aqua": "#7FDBFF",
        "black": "#111111",
        "blue": "#0074D9",
        "blueDarker": "#005daa",
        "blueLighter": "#f0fbff",
        "fuchsia": "#F012BE",
        "gray": "#AAAAAA",
        "grayDark": "#5e5e5e",
        "grayDarker": "#3a3a3a",
        "green": "#2ECC40",
        "lime": "#01FF70",
        "maroon": "#85144B",
        "navy": "#001F3F",
        "olive": "#3D9970",
        "orange": "#FF851B",
        "purple": "#B10DC9",
        "red": "#FF4136",
        "silver": "#DDDDDD",
        "teal": "#39CCCC",
        "white": "#FFFFFF",
        "yellow": "#FFDC00",
      },
      "patterns": Object {
        "transparencyBackground": Array [
          "
  background-image: linear-gradient(45deg, ",
          "#AAAAAA",
          " 25%, transparent 25%),
    linear-gradient(-45deg, ",
          "#AAAAAA",
          " 25%, transparent 25%),
    linear-gradient(45deg, transparent 75%, ",
          "#AAAAAA",
          " 75%),
    linear-gradient(-45deg, transparent 75%, ",
          "#AAAAAA",
          " 75%);
  background-color: ",
          "#5e5e5e",
          ";
  background-size: 20px 20px;
  background-position: 0 0, 0 10px, 10px -10px, -10px 0px;
",
        ],
      },
    }
  }
>
  <StyledComponent
    forwardedClass={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": undefined,
        "componentStyle": ComponentStyle {
          "componentId": "sc-eNQAEJ",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "
  flex-basis: 500px;
  margin: 8px;
  flex-shrink: 1;
  border: 1px solid ",
            [Function],
            ";
  border-radius: 4px;
  ",
            [Function],
            ";
  cursor: pointer;
  position: relative;
  display: flex;
  padding: 36px 8px 8px;
  flex-wrap: wrap;
  align-items: middle;
  justify-content: space-around;

  .product-heading {
    background: ",
            [Function],
            ";
    border-bottom: 1px solid ",
            [Function],
            ";
    text-align: center;
    padding: 4px 0;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    border-top-left-radius: 4px;
    border-top-right-radius: 4px;
  }
  .preview {
    width: 50%;
    flex-grow: 1;
    padding: 8px;
    box-sizing: border-box;
    transition: 0.6s transform;

    &.count-2 {
      width: 100%;
    }
    > svg {
      height: 100%;
      width: 100%;
    }
  }
  &:hover {
    .preview {
      transform: translateY(-15px);
    }
  }
",
          ],
        },
        "displayName": "styled.div",
        "render": [Function],
        "styledComponentId": "sc-eNQAEJ",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    theme={
      Object {
        "colors": Object {
          "aqua": "#7FDBFF",
          "black": "#111111",
          "blue": "#0074D9",
          "blueDarker": "#005daa",
          "blueLighter": "#f0fbff",
          "fuchsia": "#F012BE",
          "gray": "#AAAAAA",
          "grayDark": "#5e5e5e",
          "grayDarker": "#3a3a3a",
          "green": "#2ECC40",
          "lime": "#01FF70",
          "maroon": "#85144B",
          "navy": "#001F3F",
          "olive": "#3D9970",
          "orange": "#FF851B",
          "purple": "#B10DC9",
          "red": "#FF4136",
          "silver": "#DDDDDD",
          "teal": "#39CCCC",
          "white": "#FFFFFF",
          "yellow": "#FFDC00",
        },
        "patterns": Object {
          "transparencyBackground": Array [
            "
  background-image: linear-gradient(45deg, ",
            "#AAAAAA",
            " 25%, transparent 25%),
    linear-gradient(-45deg, ",
            "#AAAAAA",
            " 25%, transparent 25%),
    linear-gradient(45deg, transparent 75%, ",
            "#AAAAAA",
            " 75%),
    linear-gradient(-45deg, transparent 75%, ",
            "#AAAAAA",
            " 75%);
  background-color: ",
            "#5e5e5e",
            ";
  background-size: 20px 20px;
  background-position: 0 0, 0 10px, 10px -10px, -10px 0px;
",
          ],
        },
      }
    }
  >
    <div
      className="c0"
    />
  </StyledComponent>
</styled.div>
`;

exports[`ProductShowcase renders 1`] = `
<styled.div
  as={[Function]}
  to="/create/product-1"
>
  <Styled(H3)
    className="product-heading"
  >
    Krazy Kite
  </Styled(H3)>
  <Svg
    className="preview count-1"
    key="Standard"
    svg="<div>Kool Kite</div>"
  />
</styled.div>
`;
